### PR TITLE
Added strip auth option and middleware

### LINKED ISSUE #1273: Disable forwarding auth related data to the upstream
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**

Currently auth related header information is proxied to the upstream.

**What is the expected behavior?**

Auth related headers can be stripped from requests when proxied through the gateway as it presents a possible security issue.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**

All versions

### PR DESCRIPTION
**Pull request by lonelycode**
Fixes #1273

Adds a new option to the API Definition (this PR modifies the vendored version, actual change to tykcommon to follow) called "strip_auth_data" to the root of the object.

Setting this to true will cause the middleware to:

- Check if there are query string params to remove (if enabled) and remove them
- Check if there is an auth header to remove, and remove it
- If no header name is enforced (e.g. Basic Auth, HMAC or OIDC all use the "Authorization" header), strip that name instead

### CODE REVIEW

**Comment by mvdan** [comment]
Did you do this against release-2.3 on purpose? I thought we wouldn't do another 2.3.x release.

File: mw_strip_auth.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// GetConfig retrieves the configuration from the API config - we user mapstructure for this for simplicity
+func (m *MWStripAuthData) GetConfig() (interface{}, error) {
+	var thisModuleConfig MiddlewareContextVarsConfig
+	return thisModuleConfig, nil
+}
+
+func (a *MWStripAuthData) IsEnabledForSpec() bool {
+	if a.Spec.StripAuthData {
```
Comment (MangoMM) [inline mw_strip_auth.go]: `return a.Spec.StripAuthData`

**Comment by lonelycode** [comment]
Yes this is against 2.3 on purpose as it means we can just release it as a patch

**Comment by buger** [comment]
I wonder if it is more convenient to put this code directly to auth key middleware.

File: mw_strip_auth.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	key := r.Header.Get(n)
+	if key != "" {
+		r.Header.Del(n)
+	}
+}
+
+// ProcessRequest will run any checks on the request on the way through the system, return an error to have the chain fail
+func (m *MWStripAuthData) ProcessRequest(w http.ResponseWriter, r *http.Request, configuration interface{}) (error, int) {
+
+	if !m.Spec.StripAuthData {
```
Comment (MangoMM) [inline mw_strip_auth.go]: nit

```
if m.Spec.StripAuthData {
    m.StripAuth(r, mSpec)
}

return nil, 200
```

Also, does ProcessRequest function implement an interface and that's why we are also returning an error?

Comment (mvdan) [reply mw_strip_auth.go]: I wouldn't worry about nits at all, as this patch will be completely different once ported to master :)

And yes, this is an interface.

**Comment by buger** [comment]
Looks good. @ilijabojanovic ready for testing.

**Comment by buger** [comment]
@MangoMM will to help with merging to master?

**Pull request merged**